mlflow/xgboost.py

```diff
@@ -184,3 +192,89 @@ def __init__(self, xgb_model):
+def autolog(importance_types=['weight']):  # pylint: disable=W0102
+    """
+    Enables autologging from XGBoost to MLflow. Logs the following.
+
+    - parameters specified in `xgboost.train`_.
+    - metrics on each iteration (if ``evals`` specified).
+    - ``best_iteration``, ``best_ntree_limit``, and ``best_score``
+      (if ``early_stopping_rounds`` specified).
+    - feature importance.
+    - trained model.
+
+    :param importance_types: importance types to log.
+
+    """
+    import xgboost
+
+    @gorilla.patch(xgboost)
+    def train(*args, **kwargs):
+
+        def mlflow_callback(env):
+            """
+            Callback for auto-logging metrics on each iteration.
+            """
+            try_mlflow_log(mlflow.log_metrics, dict(env.evaluation_result_list),
+                           step=env.iteration)
+
+        if not mlflow.active_run():
+            try_mlflow_log(mlflow.start_run)
+            auto_end_run = True
+        else:
+            auto_end_run = False
+
+        original = gorilla.get_original_attribute(xgboost, 'train')
+        unlogged_params = ['dtrain', 'evals', 'obj', 'feval', 'evals_result',
+                           'xgb_model', 'callbacks', 'learning_rates']
+        unpacked_params = ['params']
+        log_fn_args_as_params(original, args, kwargs, unlogged_params, unpacked_params)
+
+        all_arg_names = inspect.getargspec(original)[0]  # pylint: disable=W1505
+        num_pos_args = len(args)
+
+        # checking if the 'callbacks' argument of train() is set.
+        callbacks_index = all_arg_names.index('callbacks')
+        if num_pos_args >= callbacks_index + 1:
+            tmp_list = list(args)
+            tmp_list[callbacks_index] += [mlflow_callback]
+            args = tuple(tmp_list)
+        elif 'callbacks' in kwargs and kwargs['callbacks'] is not None:
+            kwargs['callbacks'] += [mlflow_callback]
+        else:
+            kwargs['callbacks'] = [mlflow_callback]
+
+        model = original(*args, **kwargs)
+
+        # checking if the 'early_stopping_rounds' argument of train() is present.
+        early_stopping_index = all_arg_names.index('early_stopping_rounds')
+        early_stopping = (num_pos_args >= early_stopping_index + 1 or
+                          'early_stopping_rounds' in kwargs)
+
+        # if 'early_stopping_rounds' is present, the output model has
+        # 'best_score', 'best_iteration', and 'best_ntree_limit'
+        # even if early stopping didn't occur.
+        if early_stopping:
+            attrs = ['best_score', 'best_iteration', 'best_ntree_limit']
+            try_mlflow_log(mlflow.log_metrics, {attr: getattr(model, attr) for attr in attrs})
+
+        # logging feature importance as artifacts.
+        for imp_type in importance_types:
+            imp = model.get_score(importance_type=imp_type)
+            filename = 'feature_importance_{}.json'.format(imp_type)
+            filepath = os.path.join(tempfile.mkdtemp(), filename)
```

Comment:
We should make sure to remove this tempdir once we log our artifact

Reply:
will fix this